File: leetcode/7_backtracking/day26/2_40.cpp
//
// Created by Promise on 2023/10/19.
//
#include "Utils.h"


class Solution {
private:
    vector <vector<int>> result;
    vector<int> path;

    void backTracking(int sum, int target, int index, vector<bool> &used, vector<int> &candidates) {
        if (sum == target) {
            result.push_back(path);
            return;
        }
        for (int i = index; i < candidates.size() && sum + candidates[i] <= target; i++) {
            if (i > 0 && candidates[i] == candidates[i - 1] && !used[i - 1]) {
                continue;
            }
            sum += candidates[i];
            path.push_back(candidates[i]);
            used[i] = true;
            backTracking(sum, target, i + 1, used, candidates);
            sum -= candidates[i];
            path.pop_back();
            used[i] = false;
        }
    }

public:
    vector <vector<int>> combinationSum2(vector<int> &candidates, int target) {
        sort(candidates.begin(), candidates.end());
        path.clear();
        result.clear();
        vector<bool> used(candidates.size(), false);
        backTracking(0, target, 0, used, candidates);
        return result;
    }
};


int main() {
    Solution s;
    vector<int> candidates = {10, 1, 2, 7, 6, 1, 5};
    int target = 8;
    vector <vector<int>> result = s.combinationSum2(candidates, target);
    for (const auto &i: result) {
        for (auto j: i) {
            cout << j << "\t";
        }
        cout << endl;
    }
    return 0;
}

File: leetcode/7_backtracking/day26/2_40.go
package day26

import (
	"sort"
	"strconv"
)

// 给定一个候选人编号的集合 candidates 和一个目标数 target ，找出 candidates 中所有可以使数字和为 target 的组合。
//
// candidates 中的每个数字在每个组合中只能使用 一次 。
//
// 注意：解集不能包含重复的组合。
func combinationSum2(candidates []int, target int) [][]int {
	var result [][]int
	var path []int
	sort.Ints(candidates)
	maps := map[string]bool{}
	var backTracking func(candidates []int, target, sum, index int)
	backTracking = func(candidates []int, target, sum, index int) {
		if target == sum {
			s := ""
			for _, v := range path {
				s += strconv.Itoa(v)
			}
			if !maps[s] {
				temp := make([]int, len(path))
				copy(temp, path)
				result = append(result, temp)
				maps[s] = true
			}
			return
		}
		for i := index; i < len(candidates); i++ {
			if i > 0 && candidates[i] == candidates[i-1] && len(path) == 0 {
				continue
			}
			if sum+candidates[i] > target {
				return
			}
			path = append(path, candidates[i])
			sum += candidates[i]
			backTracking(candidates, target, sum, i+1)
			path = path[:len(path)-1]
			sum -= candidates[i]
		}
	}
	backTracking(candidates, target, 0, 0)
	return result
}

func T(a []int, b int) [][]int {
	return combinationSum2(a, b)
}
